packages/Chem/src/analysis/molecular-matched-pairs/mmp-generations.ts

```diff
@@ -0,0 +1,114 @@
+  const rulesTo = rulesColumns.byName('To');
+  const activityN = activityMeanNames.length;
+
+  const structures = molecules.toList();
+  const cores = new Array<Array<string>>(activityN);
+  const from = new Array<Array<string>>(activityN);
+  const to = new Array<Array<string>>(activityN);
+  const prediction = new Array<Array<number>>(activityN);
+
+  for (let i = 0; i < activityN; i ++) {
+    cores[i] = new Array<string>(structures.length);
+    from[i] = new Array<string>(structures.length);
+    to[i] = new Array<string>(structures.length);
+    prediction[i] = new Array<number>(structures.length);
+    for (let j = 0; j < structures.length; j++)
+      prediction[i][j] = 0;
+  }
+
+  for (let i = 0; i < structures.length; i ++) {
+    const singleActivities = new Array<number>(activityN);
+    for (let j = 0; j < activityN; j++)
+      singleActivities[j] = activities.byName(activityMeanNames[j].replace('Mean Difference ', '')).get(i);
+
+    for (let j = 0; j < frags[i].length; j++) {
+      const core = frags[i][j][0];
+      const subst = frags[i][j][1];
+      if (core != '') {
+        for (let k = 0; k < rulesFrom.length; k++) {
+          if (subst === rulesFrom.get(k)) {
+            for (let kk = 0; kk <activityN; kk++) {
+              const activity = singleActivities[kk] + rulesColumns.byName(activityMeanNames[kk]).get(k);
+              if (activity > prediction[kk][i]) {
+                prediction[kk][i] = activity;
+                cores[kk][i] = core;
+                from[kk][i] = subst;
+                to[kk][i] = rulesTo.get(k);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  const allStructures = Array(structures.length * activityN).fill(0).map((_, i) => structures[i % structures.length]);
+
+  const colStructure = DG.Column.fromList('string', 'structure', allStructures);
+  colStructure.semType = DG.SEMTYPE.MOLECULE;
+  const cols = [colStructure];
+  let allCores: Array<string> = [];
+  let allFrom: Array<string> = [];
+  let allTo: Array<string> = [];
+  let allInits: Array<number> = [];
+  let activityName: Array<string> = [];
+  let allPreds: Array<number> = [];
+
+  for (let i = 0; i < activityN; i ++) {
+    const name = activityMeanNames[i].replace('Mean Difference ', '');
+    allInits = allInits.concat(activities.byName(name).toList());
+    activityName = activityName.concat(Array(structures.length).fill(name));
+
+    allCores = allCores.concat(cores[i]);
+    allFrom = allFrom.concat(from[i]);
+    allTo = allTo.concat(to[i]);
+    allPreds = allPreds.concat(prediction[i]);
+  }
+
+  const generation = new Array<string>(allCores.length);
+
+  for (let i = 0; i < allCores.length; i++) {
+    let mol = null;
+    let smilesGen = '';
+    try {
+      const smi = `${allCores[i]}.${allTo[i]}`.replaceAll('([*:1])', '9').replaceAll('[*:1]', '9');
+      mol = module.get_mol(smi);
+      smilesGen = mol.get_smiles();
+      generation[i] = smilesGen;
+      mol!.delete();
```

Comment:
Move the deletion to the finally block, otherwise there is a leak if get_smiles throws (can it?)